nixos/modules/services/misc/filebrowser.nix

```
@@ -0,0 +1,152 @@
+            };
+
+            port = lib.mkOption {
+              default = 5983;
+              description = "port to listen on";
+              type = lib.types.port;
+            };
+          };
+        };
+      };
+
+      user = lib.mkOption {
+        default = "filebrowser";
+        description = "User under which Filebrowser runs";
+        type = lib.types.str;
+      };
+    };
+  };
+
+  config = let
+    cfg = config.services.filebrowser;
+
+    cmd =
+      "${cfg.package}/bin/filebrowser --config='${cfg.dataDir}/config.json' --database='${cfg.dataDir}/database.boltdb'";
+
+    # This `preamble` should be included before working with filebrowser or its data files
+    preamble = ''
+      # create the `config.json` file if it doesn't exist, and either way, make sure permissions
+      # are correct in case use of `sudo` or something mangled them
+      if [ ! -f '${cfg.dataDir}/config.json' ] ; then
+        echo "{}" > '${cfg.dataDir}/config.json' || exit $?
+      fi
+      chown ${cfg.user}:${cfg.group} '${cfg.dataDir}/config.json' || exit $?
+      chmod 600 '${cfg.dataDir}/config.json' || exit $?
+
+      # create the `database.boltdb` file if it doesn't exist, and either way, make sure permissions
+      # are correct in case use of `sudo` or something mangled them
+      if [ ! -f '${cfg.dataDir}/database.boltdb' ] ; then
+        ${cmd} config init 2>&1 >/dev/null || exit $?
+      fi
+      chown ${cfg.user}:${cfg.group} '${cfg.dataDir}/database.boltdb' || exit $?
+      chmod 600 '${cfg.dataDir}/database.boltdb' || exit $?
+    '';
+
+  in lib.mkIf cfg.enable {
+
+    # `filebrowser-service-tool` is a helper command to make it much easier to correctly
+    # update, add, and remove settings, users, rules, etc, in the filebrowser db. Eventually,
+    # I'd like this to go away once we've correctly modeled the full set of parameters.
+    environment.systemPackages = [
+      (pkgs.writeShellScriptBin "filebrowser-service-tool" ''
+        if systemctl is-active --quiet filebrowser.service ; then
+          echo "'filebrowser.service' needs to be stopped before using this tool"
+          exit 1
+        fi
+
+        ${preamble}
+
+        exec ${cmd} "$@"
+      '')
+    ];
+
+    networking.firewall =
+      lib.mkIf cfg.openFirewall { allowedTCPPorts = [ cfg.settings.port ]; };
+
+    systemd.services.filebrowser = {
+      after = [ "network.target" ];
+      description = "Filebrowser Service";
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        Group = cfg.group;
+        Restart = "on-failure";
+        Type = "simple";
+        User = cfg.user;
+
+        ExecStart = pkgs.writeShellScript "filebrowser-exec-start.sh" ''
+          ${preamble}
+
+          ${cmd} config set --address=${cfg.settings.address} || exit $?
+          ${cmd} config set --port=${
+            builtins.toString cfg.settings.port
+          } || exit $?
+
+          exec ${cmd}
+        '';
+      };
+
```

```suggestion
      };
      script = ''
        ${preamble}

        ${cmd} config set --address=${cfg.settings.address} || exit $?
        ${cmd} config set --port=${toString cfg.settings.port} || exit $?

        exec ${cmd}
      '';
```
we don't need the extra wrapper here